src/linker/Linker.Dataflow/XmlFlowAnnotationSource.cs

```diff
@@ -0,0 +1,690 @@
+		}
+
+		public DynamicallyAccessedMemberTypes GetFieldAnnotation (FieldDefinition field)
+		{
+			return _fields.TryGetValue (field, out var ann) ? ann : DynamicallyAccessedMemberTypes.None;
+		}
+
+		public DynamicallyAccessedMemberTypes GetParameterAnnotation (MethodDefinition method, int index)
+		{
+			if (_methods.TryGetValue (method, out var ann) && ann.ParameterAnnotations != null) {
+				string paramName = method.Parameters[index].Name;
+
+				foreach (var (ParamName, Annotation) in ann.ParameterAnnotations)
+					if (ParamName == paramName)
+						return Annotation;
+			}
+
+			return DynamicallyAccessedMemberTypes.None;
+		}
+
+		public DynamicallyAccessedMemberTypes GetPropertyAnnotation (PropertyDefinition property)
+		{
+			return _properties.TryGetValue (property, out var ann) ? ann : DynamicallyAccessedMemberTypes.None;
+		}
+
+		public DynamicallyAccessedMemberTypes GetReturnParameterAnnotation (MethodDefinition method)
+		{
+			return _methods.TryGetValue (method, out var ann) ? ann.ReturnAnnotation : DynamicallyAccessedMemberTypes.None;
+		}
+
+		public DynamicallyAccessedMemberTypes GetThisParameterAnnotation (MethodDefinition method)
+		{
+			if (_methods.TryGetValue (method, out var ann) && ann.ParameterAnnotations != null) {
+
+				foreach (var (ParamName, Annotation) in ann.ParameterAnnotations)
+					if (ParamName == "this")
+						return Annotation;
+			}
+
+			return DynamicallyAccessedMemberTypes.None;
+		}
+
+		static DynamicallyAccessedMemberTypes ParseKinds (ArrayBuilder<Attribute> attributes)
+		{
+			foreach (var attribute in attributes.ToArray ()) {
+				if (attribute.attributeName == "System.Runtime.CompilerServices.DynamicallyAccessedMembers" && attribute.arguments.Count == 1) {
+					foreach (var argument in attribute.arguments.ToArray ()) {
+						if (argument == string.Empty)
+							break;
+						return (DynamicallyAccessedMemberTypes) Enum.Parse (typeof (DynamicallyAccessedMemberTypes), argument);
+					}
+				}
+			}
+			return DynamicallyAccessedMemberTypes.None;
+		}
+
+		void Initialize ()
+		{
+			XPathNavigator nav = _document.CreateNavigator ();
+
+			// This step can be created with XML files that aren't necessarily
+			// annotations descriptor files. So bail if we don't have a <annotations> element.
+			if (!nav.MoveToChild ("annotations", _ns))
+				return;
+
+			try {
+				ProcessAssemblies (_context, nav.SelectChildren ("assembly", _ns));
+			} catch (Exception ex) when (!(ex is XmlResolutionException)) {
+				throw new XmlResolutionException (string.Format ("Failed to process XML description: {0}", _document), ex);
+			}
+		}
+
+		protected virtual void ProcessAssemblies (LinkContext context, XPathNodeIterator iterator)
+		{
+			while (iterator.MoveNext ()) {
+				AssemblyDefinition assembly = GetAssembly (context, GetAssemblyName (iterator.Current));
+
+				if (assembly == null) {
+					context.LogMessage ($"Assembly {GetAssemblyName (iterator.Current).Name} couldn't be resolved");
```

Comment:
Use the new logging APIs (`MessageContainer`).

Reply:
Now Im ussing the MessageContainer but had to modify error codes to make them more generic